src/ring_theory/polynomial/basic.lean

```diff
@@ -429,13 +429,76 @@ exists_irreducible_of_nat_degree_pos $ nat.pos_of_ne_zero hf
 lemma linear_independent_powers_iff_eval₂
   (f : M →ₗ[R] M) (v : M) :
   linear_independent R (λ n : ℕ, (f ^ n) v)
-    ↔ ∀ (p : polynomial R), polynomial.aeval f p v = 0 → p = 0 :=
+    ↔ ∀ (p : polynomial R), aeval f p v = 0 → p = 0 :=
 begin
   rw linear_independent_iff,
   simp only [finsupp.total_apply, aeval_endomorphism],
   refl
 end
 
+lemma disjoint_ker_aeval_of_coprime
+  (f : M →ₗ[R] M) {p q : polynomial R} (hpq : is_coprime p q) :
+  disjoint (aeval f p).ker (aeval f q).ker :=
+begin
+  intros v hv,
+  rcases hpq with ⟨p', q', hpq'⟩,
+  simpa [linear_map.mem_ker.1 (submodule.mem_inf.1 hv).1,
+         linear_map.mem_ker.1 (submodule.mem_inf.1 hv).2]
+    using congr_arg (λ p : polynomial R, aeval f p v) hpq'.symm,
+end
+
+lemma sup_aeval_range_eq_top_of_coprime
+  (f : M →ₗ[R] M) {p q : polynomial R} (hpq : is_coprime p q) :
+  (aeval f p).range ⊔ (aeval f q).range = ⊤ :=
+begin
+  rw eq_top_iff,
+  intros v hv,
+  rw submodule.mem_sup,
+  rcases hpq with ⟨p', q', hpq'⟩,
+  use aeval f (p * p') v,
+  use linear_map.mem_range.2 ⟨aeval f p' v, by simp only [linear_map.mul_app, aeval_mul]⟩,
+  use aeval f (q * q') v,
+  use linear_map.mem_range.2 ⟨aeval f q' v, by simp only [linear_map.mul_app, aeval_mul]⟩,
+  simpa only [mul_comm p p', mul_comm q q', aeval_one, aeval_add]
+    using congr_arg (λ p : polynomial R, aeval f p v) hpq'
+end
+
+lemma sup_ker_aeval_le_ker_aeval_mul
+  {f : M →ₗ[R] M} {p q : polynomial R} :
+  (aeval f p).ker ⊔ (aeval f q).ker ≤
+    (aeval f (p * q)).ker :=
+begin
+  intros v hv,
+  rcases submodule.mem_sup.1 hv with ⟨x, hx, y, hy, hxy⟩,
+  have h_eval_x : aeval f (p * q) x = 0,
+  { rw [mul_comm, aeval_mul, linear_map.mul_apply, linear_map.mem_ker.1 hx, linear_map.map_zero] },
+  have h_eval_y : aeval f (p * q) y = 0,
+  { rw [aeval_mul, linear_map.mul_apply, linear_map.mem_ker.1 hy, linear_map.map_zero] },
+  rw [linear_map.mem_ker, ←hxy, linear_map.map_add, h_eval_x, h_eval_y, add_zero],
+end
+
+lemma sup_ker_aeval_eq_ker_aeval_mul_of_coprime
+  (f : M →ₗ[R] M) {p q : polynomial R} (hpq : is_coprime p q) :
+  (aeval f p).ker ⊔ (aeval f q).ker =
+    (aeval f (p * q)).ker :=
+begin
+  apply le_antisymm sup_ker_aeval_le_ker_aeval_mul,
+  intros v hv,
+  rw submodule.mem_sup,
+  rcases hpq with ⟨p', q', hpq'⟩,
+  have h_eval₂_qpp' : aeval f (q * (p * p')) v = 0,
+  { rw [mul_comm, mul_assoc, mul_comm, mul_assoc, aeval_mul,
+      mul_comm, linear_map.mul_apply, linear_map.mem_ker.1 hv, linear_map.map_zero] },
```

Comment:
A long list of `mul_comm` and `mul_assoc` is hard to follow, maybe you can use `calc` or `suffices _, { ring, assumption }` instead?

Reply:
It seems `ring` doesn't realize it needs to work on the subterm containing the multiplication? And it is very slow! So I kept the `rw`, but used `calc` to clarify what is going on.